### PR TITLE
Packages: Finish the constants package

### PR DESCRIPTION
**Pull request by tyxla**
This PR is an attempt to make the constants package a properly namespaced package with independent tests, and to have the rest of Jetpack use that package the new way, without breaking Jetpack or the existing tests.

#### Changes proposed in this Pull Request:
* Add a namespace to the constants package.
* Update the constants package to use PSR-4.
* Fix all the phpcs errors in the constants package and add some inline docs.
* Add a constants package README with examples.
* Move the existing constant tests to the package.
* Add the infrastructure to automatically test the constants package.
* Update the rest of Jetpack to use the new constants package.
* Update composer.lock with the new changes.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Part of the Jetpack DNA project - p1HpG7-70O-p2

#### Testing instructions:
* Checkout this branch.
* Run `rm -rf ./vendor` inside Jetpack.
* Run `composer install` inside Jetpack.
* Navigate to `packages/constants`
* Run `composer phpunit`.
* Verify that runs `composer install` internally, and then runs the package tests.
* Verify the package tests pass.
* Run `composer phpunit` again.
* Verify the second time is fast because the dependencies are already installed.
* Verify CI passes on this PR and 100% is green.
* Smoke test Jetpack to verify it still works well in both the admin and the frontend.

#### Proposed changelog entry for your changes:
* Finish the constants package and make its tests independent.

### CODE REVIEW

**Comment by tyxla** [comment]
This PR is ready for review @Automattic/jetpack-crew @Automattic/poseidon @gravityrail

**Review by roccotripaldi — COMMENTED** [review]
Looks great! It looks like fusion struggled making a Phabricator changeset for this PR, and we'll need to figure out how to ensure we to autoload on .com so that fusion continues to work.

**Comment by tyxla** [comment]
This one has been rebased and is ready for a review. 

> It looks like fusion struggled making a Phabricator changeset for this PR

Yeah, that would happen when we're touching files that are not Fusion-ed in this PR. 

Anyone from @Automattic/jetpack-crew would you recommend that I split this to 2 PRs - one that touches only Fusion-ed files, and one that touches only ones that aren't Fusion-ed?

> and we'll need to figure out how to ensure we to autoload on .com so that fusion continues to work.

Yeah, I imagine we'll end up having to take care of that regardless, since we're introducing various packages and they'll end up being used on the .com codebase as well. What do you think?

**Comment by jeherve** [comment]
> Anyone from @Automattic/jetpack-crew would you recommend that I split this to 2 PRs - one that touches only Fusion-ed files, and one that touches only ones that aren't Fusion-ed?

I think this is part of a larger discussion we should have about syncing packages. We'll run into more and more as we go, and we need to find a good way to ensure packages are well synced. One way could be to actually rely on the packages on WordPress.com, for example.

I don't think that's something you need to worry about in this PR since we are not merging into `master` just yet, but I think that's a discussion we should start on +jetpackdna.

What do you think?

**Review by jeherve — APPROVED** [review]
This tests well on my end. Should be good to merge!

File: packages/constants/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+Testing constants is hard. Once you define a constant in PHP, it's defined. Constants Manager is an abstraction layer so that unit tests can set constants for tests.
+
+### Usage
+
+Retrieve the value of a constant `CONSTANT_NAME` (returns `null` if it's not defined):
+
+```php
+use Automattic\Jetpack\Constants\Manager as Constants_Manager;
+
+$constant_name = Constants_Manager::get_constant( 'CONSTANT_NAME' );
```
Comment (lezama) [inline packages/constants/README.md]:
```suggestion
$constant_value = Constants_Manager::get_constant( 'CONSTANT_NAME' );
```

File: packages/constants/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+Set the value of a constant `CONSTANT_NAME` to a partcular value:
```
Comment (lezama) [inline packages/constants/README.md]:
```suggestion
Set the value of a constant `CONSTANT_NAME` to a particular value:
```

File: packages/constants/src/Manager.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * A constant manager for Jetpack.
```
Comment (lezama) [inline packages/constants/src/Manager.php]:
```suggestion
 * A constants manager for Jetpack.
```

**Comment by tyxla** [comment]
Thanks @lezama and @jeherve! Going to merge when tests are green.

**Comment by tyxla** [comment]
> > Anyone from @Automattic/jetpack-crew would you recommend that I split this to 2 PRs - one that touches only Fusion-ed files, and one that touches only ones that aren't Fusion-ed?
> 
> I think this is part of a larger discussion we should have about syncing packages. We'll run into more and more as we go, and we need to find a good way to ensure packages are well synced. One way could be to actually rely on the packages on WordPress.com, for example.
> 
> I don't think that's something you need to worry about in this PR since we are not merging into `master` just yet, but I think that's a discussion we should start on +jetpackdna.
> 
> What do you think?

You're totally right! This will be a recurring problem, so let's discuss together and figure out the best way forward.

**Pull request merged**